File: packages/dotnet/DeRec.Library/src/ContactMessage.cs
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 DeRec Alliance. All rights reserved.

using Google.Protobuf;

namespace DeRec.Library;

/// <summary>
/// Decoded representation of a DeRec <c>ContactMessage</c>, exchanged out-of-band
/// before pairing begins.
/// </summary>
/// <param name="ChannelId">
/// Channel identifier the initiator expects to use for the new pairing session.
/// </param>
/// <param name="ContactMode">
/// Selects how the public encryption material is delivered. See
/// <see cref="ContactMode"/>.
/// </param>
/// <param name="TransportProtocol">
/// Transport endpoint and protocol to use when sending protocol messages to the initiator.
///
/// <para>
/// <b>Reading this directly is incorrect.</b> Its meaning narrowed from "the
/// endpoint" to "one entry of a list, and possibly absent": a creator that has
/// moved past this field populates only <see cref="SupportedTransports"/>, and
/// this property is then an empty-URI placeholder. Call
/// <see cref="AdvertisedEndpoints"/>, which resolves both spellings. Removed at
/// 0.0.5.
/// </para>
/// </param>
/// <param name="Nonce">
/// Random nonce that binds the pairing request to this contact exchange.
/// </param>
/// <param name="MlkemEncapsulationKey">
/// Serialized ML-KEM-768 encapsulation key. Present only when
/// <see cref="ContactMode"/> is <see cref="ContactMode.InlineKeys"/>.
/// </param>
/// <param name="EciesPublicKey">
/// Serialized ECIES public key. Present only when
/// <see cref="ContactMode"/> is <see cref="ContactMode.InlineKeys"/>.
/// </param>
/// <param name="ContactBindingHash">
/// SHA-384 commitment to the public encryption material. Present only when
/// <see cref="ContactMode"/> is <see cref="ContactMode.HashedKeys"/>. Validated
/// by the scanner after it receives the keys via <c>PrePair</c>.
/// </param>
public sealed record ContactMessage(
    ulong ChannelId,
    ContactMode ContactMode,
    TransportProtocol TransportProtocol,
    ulong Nonce,
    byte[]? MlkemEncapsulationKey,
    byte[]? EciesPublicKey,
    byte[]? ContactBindingHash
)
{
    /// <summary>
    /// Every transport endpoint the creator of this contact can be reached on,
    /// in its own preference order.
    /// </summary>
    /// <remarks>
    /// Empty means "only <see cref="TransportProtocol"/> is offered", which is
    /// how every implementation predating this field advertises. Preserved
    /// across a decode/encode round trip, so a contact this SDK parses and
    /// hands back to <c>Pairing.Request.Produce</c> still advertises every
    /// endpoint the creator offered.
    /// </remarks>
    public IReadOnlyList<TransportProtocol> SupportedTransports { get; init; } =
        Array.Empty<TransportProtocol>();

    /// <summary>
    /// The endpoints this contact advertises, in the creator's own order.
    /// </summary>
    /// <remarks>
    /// Yields <see cref="SupportedTransports"/> when it is non-empty, and
    /// otherwise the singular <see cref="TransportProtocol"/> — which is how
    /// every implementation predating the offer list advertises, and the reason
    /// this is a method rather than a property read. Reports what was
    /// advertised, not what is acceptable; nothing here is validated.
    /// </remarks>
    public IReadOnlyList<TransportProtocol> AdvertisedEndpoints() =>
        SupportedTransports.Count > 0
            ? SupportedTransports
            : string.IsNullOrEmpty(TransportProtocol.Uri)
                ? Array.Empty<TransportProtocol>()
                : new[] { TransportProtocol };

    /// <summary>
    /// Serializes this <see cref="ContactMessage"/> to protobuf wire bytes.
    /// </summary>
    /// <remarks>
    /// Structurally validates the contact's <c>(ContactMode, inline keys,
    /// binding hash)</c> tuple before emitting bytes — a locally constructed
    /// instance that violates the per-mode invariant raises
    /// <see cref="DeRecException"/> instead of producing a wire blob that
    /// downstream consumers would reject anyway.
    /// </remarks>
    // Touches the deprecated singular `transportProtocol`: this is the
    // compatibility path that keeps peers predating `supportedTransports`
    // working, so the warning is expected here rather than a defect.
#pragma warning disable CS0612
    internal byte[] ToProtoBytes()
    {
        var proto = new Org.Derecalliance.Derec.Protobuf.ContactMessage
        {
            ChannelId = ChannelId,
            ContactMode = (Org.Derecalliance.Derec.Protobuf.ContactMode)(int)ContactMode,
            TransportProtocol = new Org.Derecalliance.Derec.Protobuf.TransportProtocol
            {
                Uri = TransportProtocol.Uri,
                Protocol = (Org.Derecalliance.Derec.Protobuf.Protocol)(int)TransportProtocol.Protocol,
            },
            Nonce = Nonce,
        };
        foreach (TransportProtocol offer in SupportedTransports)
        {
            proto.SupportedTransports.Add(new Org.Derecalliance.Derec.Protobuf.TransportProtocol
            {
                Uri = offer.Uri,
                Protocol = (Org.Derecalliance.Derec.Protobuf.Protocol)(int)offer.Protocol,
            });
        }
        if (MlkemEncapsulationKey is { Length: > 0 } mlkem)
        {
            proto.MlkemEncapsulationKey = Google.Protobuf.ByteString.CopyFrom(mlkem);
        }
        if (EciesPublicKey is { Length: > 0 } ecies)
        {
            proto.EciesPublicKey = Google.Protobuf.ByteString.CopyFrom(ecies);
        }
        if (ContactBindingHash is { Length: > 0 } hash)
        {
            proto.ContactBindingHash = Google.Protobuf.ByteString.CopyFrom(hash);
        }
        byte[] bytes = proto.ToByteArray();
        Validate(bytes);
        return bytes;
    }

    /// <summary>
    /// Deserializes a <see cref="ContactMessage"/> from protobuf wire bytes
    /// and structurally validates the result against the per-mode invariant
    /// documented on the wire format. Throws <see cref="DeRecException"/>
    /// if the contact is malformed (unknown <see cref="ContactMode"/>,
    /// mode/field mismatch, wrong binding-hash length).
    /// </summary>
    // Touches the deprecated singular `transportProtocol`: this is the
    // compatibility path that keeps peers predating `supportedTransports`
    // working, so the warning is expected here rather than a defect.
#pragma warning disable CS0612
    internal static ContactMessage FromProtoBytes(byte[] bytes)
    {
        Validate(bytes);

        var proto = Org.Derecalliance.Derec.Protobuf.ContactMessage.Parser.ParseFrom(bytes);

        var tp = proto.TransportProtocol is { } protoTp
            ? new TransportProtocol(protoTp.Uri, (Protocol)(int)protoTp.Protocol)
            : new TransportProtocol(string.Empty);

        // proto3 `optional bytes` fields are reported via `HasFoo` once set;
        // if the field was never set the property still returns `ByteString.Empty`,
        // so we map that case to `null` to match the wire semantics.
        byte[]? mlkem = proto.HasMlkemEncapsulationKey
            ? proto.MlkemEncapsulationKey.ToByteArray()
            : null;
        byte[]? ecies = proto.HasEciesPublicKey
            ? proto.EciesPublicKey.ToByteArray()
            : null;
        byte[]? hash = proto.HasContactBindingHash
            ? proto.ContactBindingHash.ToByteArray()
            : null;

        return new ContactMessage(
            ChannelId: proto.ChannelId,
            ContactMode: (ContactMode)(int)proto.ContactMode,
            TransportProtocol: tp,
            Nonce: proto.Nonce,
            MlkemEncapsulationKey: mlkem,
            EciesPublicKey: ecies,
            ContactBindingHash: hash
        )
        {
            SupportedTransports = proto.SupportedTransports
                .Select(t => new TransportProtocol(t.Uri, (Protocol)(int)t.Protocol))
                .ToList(),
        };
    }

    private static void Validate(byte[] bytes)
    {
        Native.DeRecError error =
            Native.Pairing.validate_contact_message(bytes, (UIntPtr)bytes.Length);
        Utils.ThrowIfError(error);
    }
}
#pragma warning restore CS0612

File: library/src/extensions/contact_message.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 DeRec Alliance. All rights reserved.

use crate::derec_message::current_timestamp;
use crate::primitives::pairing::PairingError;
use crate::types::ChannelId;
use derec_cryptography::pairing::PairingContactMessageMaterial;
use derec_proto::{ContactMessage, ContactMode, TransportProtocol};

/// Extension trait that attaches structural validation as a method on
/// the proto [`ContactMessage`].
///
/// The proto type lives in [`derec_proto`], so an inherent `impl` is
/// impossible from this crate — this trait sidesteps the orphan rule
/// the same way [`TransportProtocolExt`](super::transport_protocol::TransportProtocolExt) does for
/// [`derec_proto::TransportProtocol`]. Bring the trait into scope
/// (`use crate::extensions::contact_message::ContactMessageExt as _;`) at every parse
/// boundary that ingests an untrusted `ContactMessage` and call
/// `contact.validate()?` before handing it to protocol code.
pub(crate) trait ContactMessageExt {
    /// Structural validator for a decoded [`ContactMessage`]. Enforces
    /// the per-mode field-presence invariants the proto schema
    /// documents but cannot itself express:
    ///
    /// - The declared `contact_mode` must be a known [`ContactMode`]
    ///   value.
    /// - [`ContactMode::InlineKeys`]: `mlkem_encapsulation_key` and
    ///   `ecies_public_key` MUST be present and non-empty;
    ///   `contact_binding_hash` MUST be absent (would mean the contact
    ///   is misshaped — keys present AND a commitment).
    /// - [`ContactMode::HashedKeys`]: both inline-key fields MUST be
    ///   absent, and `contact_binding_hash` MUST be present with
    ///   exactly 48 bytes (SHA-384 digest size).
    /// - [`ContactMode::NoKeys`]: all three of
    ///   `mlkem_encapsulation_key`, `ecies_public_key`, and
    ///   `contact_binding_hash` MUST be absent. The contact carries
    ///   only `channel_id`, `nonce`, and `transport_protocol`; any
    ///   key-shaped field on a NoKeys contact is a protocol misuse.
    ///
    /// The bidirectional check matters most for `HashedKeys` and
    /// `NoKeys`: a malformed `HashedKeys` contact that carried inline
    /// keys could have its keys consumed downstream without the
    /// binding-hash recomputation that gives the mode its MITM
    /// resistance; a malformed `NoKeys` contact that carried
    /// unauthenticated key material would silently upgrade the flow to
    /// a mode the OOB channel wasn't trusted for.
    fn validate(&self) -> Result<(), crate::Error>;

    /// Returns `true` when this contact's mode requires a `PrePairRequest`
    /// round-trip before the responder can produce a `PairRequestMessage`.
    ///
    /// - [`ContactMode::HashedKeys`] and [`ContactMode::NoKeys`] return
    ///   `true` — both require the responder to fetch key material via
    ///   `PrePair` before the encrypted pairing handshake can begin.
    /// - [`ContactMode::InlineKeys`] returns `false` — the keys are
    ///   already carried in the contact itself.
    /// - Unknown enum values return `false`; downstream validation
    ///   (e.g. [`Self::validate`]) is responsible for rejecting them.
    fn requires_pre_pair(&self) -> bool;

    /// Build a [`ContactMode::InlineKeys`] [`ContactMessage`] carrying
    /// the given ML-KEM + ECIES public key material verbatim.
    ///
    /// `own` is every endpoint the contact creator serves, in its own
    /// preference order. It fills `supported_transports` wholesale, and
    /// its first entry also fills the legacy singular
    /// `transport_protocol` so peers predating the offer list still find
    /// an endpoint. Deriving the singular field here rather than taking
    /// it as a second argument is what keeps the two from disagreeing.
    /// An empty `own` yields an absent singular field; callers reject
    /// that case before constructing a contact.
    ///
    /// Timestamp is stamped with the current wall-clock; callers that
    /// need deterministic timestamps must construct the proto struct
    /// directly.
    fn inline_keys(
        channel_id: ChannelId,
        nonce: u64,
        own: Vec<TransportProtocol>,
        pk: PairingContactMessageMaterial,
    ) -> ContactMessage;

    /// Build a [`ContactMode::HashedKeys`] [`ContactMessage`]: no key
    /// material inlined, but a SHA-384 commitment (`contact_binding_hash`)
    /// over `(mlkem_encapsulation_key || ecies_public_key || u64_be(nonce)
    /// || u64_be(channel_id))` so the scanner can verify keys received
    /// later via `PrePair` against the commitment.
    ///
    /// `own` is outside the binding hash, exactly like the singular
    /// field derived from it.
    ///
    /// Timestamp is stamped with the current wall-clock.
    fn hashed_keys(
        channel_id: ChannelId,
        nonce: u64,
        own: Vec<TransportProtocol>,
        pk: &PairingContactMessageMaterial,
    ) -> ContactMessage;

    /// Build a [`ContactMode::NoKeys`] [`ContactMessage`]: no key
    /// material and no commitment — only `channel_id`, `nonce`, and
    /// `transport_protocol`. The contact creator generates key material
    /// on the fly when the corresponding `PrePairRequest` arrives; the
    /// scanner accepts the returned keys without cryptographic
    /// verification.
    ///
    /// Trust rests entirely on the out-of-band delivery channel being
    /// fully trusted. Applications MUST rate-limit inbound
    /// `PrePairRequest`s per channel and expire outstanding NoKeys
    /// contacts on a short timer.
    ///
    /// Timestamp is stamped with the current wall-clock.
    fn no_keys(channel_id: ChannelId, nonce: u64, own: Vec<TransportProtocol>) -> ContactMessage;
}

impl ContactMessageExt for ContactMessage {
    // Compatibility, not oversight — see the `transport` module docs.
    #[allow(deprecated)]
    fn validate(&self) -> Result<(), crate::Error> {
        let mode = ContactMode::try_from(self.contact_mode).map_err(|_| {
            #[cfg(feature = "logging")]
            tracing::warn!(
                contact_mode = self.contact_mode,
                "unknown contact_mode value"
            );
            PairingError::InvalidContactMessage("unknown contact_mode value")
        })?;

        let mlkem_present = self
            .mlkem_encapsulation_key
            .as_ref()
            .is_some_and(|v| !v.is_empty());
        let ecies_present = self
            .ecies_public_key
            .as_ref()
            .is_some_and(|v| !v.is_empty());
        let hash_present = self
            .contact_binding_hash
            .as_ref()
            .is_some_and(|v| !v.is_empty());

        match mode {
            ContactMode::InlineKeys => {
                if !mlkem_present {
                    return Err(PairingError::InvalidContactMessage(
                        "inline_keys contact missing mlkem_encapsulation_key",
                    )
                    .into());
                }
                if !ecies_present {
                    return Err(PairingError::InvalidContactMessage(
                        "inline_keys contact missing ecies_public_key",
                    )
                    .into());
                }
                if hash_present {
                    return Err(PairingError::InvalidContactMessage(
                        "inline_keys contact must not carry contact_binding_hash",
                    )
                    .into());
                }
            }
            ContactMode::HashedKeys => {
                if mlkem_present || ecies_present {
                    return Err(PairingError::InvalidContactMessage(
                        "hashed_keys contact must not carry inline keys",
                    )
                    .into());
                }
                const BINDING_HASH_LEN: usize = 48;
                let hash = self.contact_binding_hash.as_ref().ok_or(
                    PairingError::InvalidContactMessage(
                        "hashed_keys contact missing contact_binding_hash",
                    ),
                )?;
                if hash.is_empty() {
                    return Err(PairingError::InvalidContactMessage(
                        "hashed_keys contact missing contact_binding_hash",
                    )
                    .into());
                }
                if hash.len() != BINDING_HASH_LEN {
                    return Err(PairingError::InvalidContactMessage(
                        "hashed_keys contact_binding_hash is not a SHA-384 digest",
                    )
                    .into());
                }
            }
            ContactMode::NoKeys => {
                if mlkem_present || ecies_present {
                    return Err(PairingError::InvalidContactMessage(
                        "no_keys contact must not carry inline keys",
                    )
                    .into());
                }
                if hash_present {
                    return Err(PairingError::InvalidContactMessage(
                        "no_keys contact must not carry contact_binding_hash",
                    )
                    .into());
                }
            }
        }

        // A contact naming no endpoint at all gives the scanner nowhere to
        // send the pair request. Either spelling satisfies this: the list,
        // or the deprecated singular field a peer predating it fills.
        let has_offers = !self.supported_transports.is_empty();
        let has_singular = self
            .transport_protocol
            .as_ref()
            .is_some_and(|tp| !tp.uri.trim().is_empty());

        if !has_offers && !has_singular {
            #[cfg(feature = "logging")]
            tracing::warn!("contact advertises no usable transport endpoint");

            return Err(PairingError::EmptyTransportUri.into());
        }

        Ok(())
    }

    fn requires_pre_pair(&self) -> bool {
        self.contact_mode == ContactMode::HashedKeys as i32
            || self.contact_mode == ContactMode::NoKeys as i32
    }

    // Compatibility, not oversight — see the `transport` module docs.
    #[allow(deprecated)]
    fn inline_keys(
        channel_id: ChannelId,
        nonce: u64,
        own: Vec<TransportProtocol>,
        pk: PairingContactMessageMaterial,
    ) -> ContactMessage {
        ContactMessage {
            channel_id: channel_id.into(),
            transport_protocol: own.first().cloned(),
            contact_mode: ContactMode::InlineKeys as i32,
            mlkem_encapsulation_key: Some(pk.mlkem_encapsulation_key),
            ecies_public_key: Some(pk.ecies_public_key),
            contact_binding_hash: None,
            nonce,
            timestamp: Some(current_timestamp()),
            supported_transports: own,
        }
    }

    // Compatibility, not oversight — see the `transport` module docs.
    #[allow(deprecated)]
    fn hashed_keys(
        channel_id: ChannelId,
        nonce: u64,
        own: Vec<TransportProtocol>,
        pk: &PairingContactMessageMaterial,
    ) -> ContactMessage {
        let binding_hash = derec_cryptography::pairing::contact_binding_hash(
            &pk.mlkem_encapsulation_key,
            &pk.ecies_public_key,
            nonce,
            channel_id.into(),
        );

        ContactMessage {
            channel_id: channel_id.into(),
            transport_protocol: own.first().cloned(),
            contact_mode: ContactMode::HashedKeys as i32,
            mlkem_encapsulation_key: None,
            ecies_public_key: None,
            contact_binding_hash: Some(binding_hash.to_vec()),
            nonce,
            timestamp: Some(current_timestamp()),
            supported_transports: own,
        }
    }

    // Compatibility, not oversight — see the `transport` module docs.
    #[allow(deprecated)]
    fn no_keys(channel_id: ChannelId, nonce: u64, own: Vec<TransportProtocol>) -> ContactMessage {
        ContactMessage {
            channel_id: channel_id.into(),
            transport_protocol: own.first().cloned(),
            contact_mode: ContactMode::NoKeys as i32,
            mlkem_encapsulation_key: None,
            ecies_public_key: None,
            contact_binding_hash: None,
            nonce,
            timestamp: Some(current_timestamp()),
            supported_transports: own,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::Error;
    use derec_proto::Protocol;
    use prost_types::Timestamp;

    fn ts(seconds: i64) -> Timestamp {
        Timestamp { seconds, nanos: 0 }
    }

    // Compatibility, not oversight — see the `transport` module docs.
    #[allow(deprecated)]
    fn well_formed_inline_keys_contact() -> ContactMessage {
        ContactMessage {
            channel_id: 42,
            transport_protocol: Some(TransportProtocol {
                uri: "https://relay.example/alice".to_owned(),
                protocol: Protocol::Https.into(),
            }),
            contact_mode: ContactMode::InlineKeys as i32,
            mlkem_encapsulation_key: Some(vec![1; 1184]),
            ecies_public_key: Some(vec![2; 33]),
            contact_binding_hash: None,
            nonce: 0xCAFE_BABE,
            timestamp: Some(ts(1_700_000_000)),
            supported_transports: Vec::new(),
        }
    }

    // Compatibility, not oversight — see the `transport` module docs.
    #[allow(deprecated)]
    fn well_formed_hashed_keys_contact() -> ContactMessage {
        ContactMessage {
            channel_id: 42,
            transport_protocol: Some(TransportProtocol {
                uri: "https://relay.example/alice/ephemeral".to_owned(),
                protocol: Protocol::Https.into(),
            }),
            contact_mode: ContactMode::HashedKeys as i32,
            mlkem_encapsulation_key: None,
            ecies_public_key: None,
            contact_binding_hash: Some(vec![0xAB; 48]),
            nonce: 0xDEAD_BEEF,
            timestamp: Some(ts(1_700_000_000)),
            supported_transports: Vec::new(),
        }
    }

    // Compatibility, not oversight — see the `transport` module docs.
    #[allow(deprecated)]
    fn well_formed_no_keys_contact() -> ContactMessage {
        ContactMessage {
            channel_id: 1234,
            transport_protocol: Some(TransportProtocol {
                uri: "https://institution.example/pair".to_owned(),
                protocol: Protocol::Https.into(),
            }),
            contact_mode: ContactMode::NoKeys as i32,
            mlkem_encapsulation_key: None,
            ecies_public_key: None,
            contact_binding_hash: None,
            nonce: 4321,
            timestamp: Some(ts(1_700_000_000)),
            supported_transports: Vec::new(),
        }
    }

    fn assert_invalid_contact(result: Result<(), crate::Error>, expected: &'static str) {
        match result {
            Err(Error::Pairing(PairingError::InvalidContactMessage(m))) => {
                assert_eq!(m, expected, "wrong InvalidContactMessage payload")
            }
            other => panic!("expected InvalidContactMessage({expected:?}), got {other:?}"),
        }
    }

    #[test]
    fn validate_accepts_well_formed_inline_keys() {
        well_formed_inline_keys_contact()
            .validate()
            .expect("well-formed inline_keys must pass");
    }

    #[test]
    fn validate_accepts_well_formed_hashed_keys() {
        well_formed_hashed_keys_contact()
            .validate()
            .expect("well-formed hashed_keys must pass");
    }

    #[test]
    fn validate_accepts_well_formed_no_keys() {
        well_formed_no_keys_contact()
            .validate()
            .expect("well-formed no_keys must pass");
    }

    #[test]
    fn validate_rejects_unknown_contact_mode() {
        let mut c = well_formed_inline_keys_contact();
        c.contact_mode = 7;
        assert_invalid_contact(c.validate(), "unknown contact_mode value");
    }

    #[test]
    fn validate_rejects_inline_missing_mlkem_key() {
        let mut c = well_formed_inline_keys_contact();
        c.mlkem_encapsulation_key = None;
        assert_invalid_contact(
            c.validate(),
            "inline_keys contact missing mlkem_encapsulation_key",
        );
    }

    #[test]
    fn validate_rejects_inline_missing_ecies_key() {
        let mut c = well_formed_inline_keys_contact();
        c.ecies_public_key = None;
        assert_invalid_contact(c.validate(), "inline_keys contact missing ecies_public_key");
    }

    #[test]
    fn validate_rejects_inline_with_binding_hash() {
        let mut c = well_formed_inline_keys_contact();
        c.contact_binding_hash = Some(vec![0xCD; 48]);
        assert_invalid_contact(
            c.validate(),
            "inline_keys contact must not carry contact_binding_hash",
        );
    }

    #[test]
    fn validate_rejects_hashed_with_inline_keys() {
        let mut c = well_formed_hashed_keys_contact();
        c.mlkem_encapsulation_key = Some(vec![1; 1184]);
        assert_invalid_contact(
            c.validate(),
            "hashed_keys contact must not carry inline keys",
        );
    }

    #[test]
    fn validate_rejects_hashed_missing_binding_hash() {
        let mut c = well_formed_hashed_keys_contact();
        c.contact_binding_hash = None;
        assert_invalid_contact(
            c.validate(),
            "hashed_keys contact missing contact_binding_hash",
        );
    }

    #[test]
    fn validate_rejects_hashed_empty_binding_hash() {
        let mut c = well_formed_hashed_keys_contact();
        c.contact_binding_hash = Some(Vec::new());
        assert_invalid_contact(
            c.validate(),
            "hashed_keys contact missing contact_binding_hash",
        );
    }

    #[test]
    fn validate_rejects_hashed_wrong_hash_length() {
        let mut c = well_formed_hashed_keys_contact();
        c.contact_binding_hash = Some(vec![0xAB; 32]);
        assert_invalid_contact(
            c.validate(),
            "hashed_keys contact_binding_hash is not a SHA-384 digest",
        );
    }

    #[test]
    fn validate_rejects_no_keys_with_mlkem_key() {
        let mut c = well_formed_no_keys_contact();
        c.mlkem_encapsulation_key = Some(vec![1; 1184]);
        assert_invalid_contact(c.validate(), "no_keys contact must not carry inline keys");
    }

    #[test]
    fn validate_rejects_no_keys_with_ecies_key() {
        let mut c = well_formed_no_keys_contact();
        c.ecies_public_key = Some(vec![2; 33]);
        assert_invalid_contact(c.validate(), "no_keys contact must not carry inline keys");
    }

    #[test]
    fn validate_rejects_no_keys_with_binding_hash() {
        let mut c = well_formed_no_keys_contact();
        c.contact_binding_hash = Some(vec![0xAB; 48]);
        assert_invalid_contact(
            c.validate(),
            "no_keys contact must not carry contact_binding_hash",
        );
    }
}
